pkg/virt-operator/kubevirt.go

```diff
@@ -0,0 +1,235 @@
+	recorder         record.EventRecorder
+	config           util.KubeVirtDeploymentConfig
+}
+
+func NewKubeVirtController(
+	clientset kubecli.KubevirtClient,
+	informer cache.SharedIndexInformer,
+	recorder record.EventRecorder) *KubeVirtController {
+
+	c := KubeVirtController{
+		clientset:        clientset,
+		queue:            workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		kubeVirtInformer: informer,
+		recorder:         recorder,
+		config:           util.GetConfig(),
+	}
+
+	c.kubeVirtInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addKubeVirt,
+		DeleteFunc: c.deleteKubeVirt,
+		UpdateFunc: c.updateKubeVirt,
+	})
+
+	return &c
+}
+
+func (c *KubeVirtController) addKubeVirt(obj interface{}) {
+	c.enqueueKubeVirt(obj)
+}
+
+func (c *KubeVirtController) deleteKubeVirt(obj interface{}) {
+	c.enqueueKubeVirt(obj)
+}
+
+func (c *KubeVirtController) updateKubeVirt(old, curr interface{}) {
+	c.enqueueKubeVirt(curr)
+}
+
+func (c *KubeVirtController) enqueueKubeVirt(obj interface{}) {
+	logger := log.Log
+	kv := obj.(*v1.KubeVirt)
+	key, err := controller.KeyFunc(kv)
+	if err != nil {
+		logger.Object(kv).Reason(err).Error("Failed to extract key from KubeVirt.")
+	}
+	c.queue.Add(key)
+}
+
+func (c *KubeVirtController) Run(threadiness int, stopCh chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.queue.ShutDown()
+	log.Log.Info("Starting KubeVirt controller.")
+
+	// Wait for cache sync before we start the pod controller
+	cache.WaitForCacheSync(stopCh, c.kubeVirtInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping KubeVirt controller.")
+}
+
+func (c *KubeVirtController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *KubeVirtController) Execute() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing KubeVirt %v", key)
+		c.queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed KubeVirt %v", key)
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *KubeVirtController) execute(key string) error {
+
+	// Fetch the latest KubeVirt from cache
+	obj, exists, err := c.kubeVirtInformer.GetStore().GetByKey(key)
+
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		// when the resource is gone, deletion was hanlded already
+		log.Log.Infof("KubeVirt resource not found")
+		return nil
+	}
+
+	kv := obj.(*v1.KubeVirt)
+	logger := log.Log.Object(kv)
+
+	if kv.DeletionTimestamp != nil {
+
+		if isKubeVirtDeleting(kv) {
+			logger.Infof("KubeVirt deletion is/was already processed")
+			return nil
+		}
+
+		log.Log.Infof("Handling deleted KubeVirt object %+v", kv)
+		err = deletion.Delete(kv, c.clientset)
+		if err != nil {
+			err1 := util.UpdatePhase(kv, v1.KubeVirtPhaseDeletionFailed, c.clientset)
```

Comment:
here's some more to clarify all the comments I'm making about phases vs conditions.

To me, these phases make sense
**Phase:** unset, deploying, deployed, deleting, deleted.

And these conditions make sense 
**Condition:** deployed failed, deleting failed 

Since a failed deploy and delete are not final phases, they seem more like a condition rather than a phase. A condition can be cleared if we eventually succeed on a subsequent iteration of the work queue.

Reply:
Exactly what I wanted to suggest! :+1: